content/courses/program-security/account-data-matching.md

```diff
@@ -427,107 +323,88 @@ pub struct SecureWithdraw<'info> {
 }
 ```
 
-#### 4. Test `secure_withdraw` instruction
+The key difference here is the use of `has_one` constraints in the
+`SecureWithdraw` struct. These ensure that the `authority`, `token_account`, and
+`withdraw_destination` passed to the instruction match those stored in the
+`Vault` account.
 
-Now let’s test the `secure_withdraw` instruction with two tests: one that uses
-`walletFake` as the authority and one that uses `wallet` as the authority. We
-expect the first invocation to return an error and the second to succeed.
+### 5. Test the Secure Withdraw Instruction
+
+Now, let's test our secure instruction:
 
 ```typescript
-describe("account-data-matching", () => {
-  ...
-  it("Secure withdraw, expect error", async () => {
-    try {
-      const tx = await program.methods
-        .secureWithdraw()
-        .accounts({
-          vault: vaultPDA,
-          tokenAccount: tokenPDA,
-          withdrawDestination: withdrawDestinationFake,
-          authority: walletFake.publicKey,
-        })
-        .transaction()
-
-      await anchor.web3.sendAndConfirmTransaction(connection, tx, [walletFake])
-    } catch (err) {
-      expect(err)
-      console.log(err)
-    }
-  })
-
-  it("Secure withdraw", async () => {
-    await spl.mintTo(
-      connection,
-      wallet.payer,
-      mint,
-      tokenPDA,
-      wallet.payer,
-      100
-    )
+it("Secure withdraw prevents unauthorized access", async () => {
+  // Setup: Initialize vault and mint some tokens to it
+  // ... (initialization code here)
 
+  // Attempt unauthorized withdrawal
+  // Note: walletFake represents an unauthorized wallet trying to withdraw funds
```

Comment:
This is one of those comments that indicate you have found a better name for something.

Rename `walletFake` to `unauthorizedWallet`

Rename `withdrawDestinationFake` to `unauthorizedWalletWithdrawDestination` or similar.

Reply:
Understood!